### PR TITLE
Adding token hash in DB for user-icu identification from short string

### PR DESCRIPTION
**Pull request by olivierteboul**
- added an authenticator class to deal with authentication from tokens
- backward compatible
- updated url generation
- add the option to have the token valid for some days. When a token is stale it is changed right before being sent to the user (and only at that moment to avoid mismatch between what the information available to the user and our db).
- will be useful for telegram support

### CODE REVIEW

File: icubam/db/store.py [hunk 1]
```diff
@@ -544,6 +600,10 @@ def get_password_hash(self, password: str) -> str:
       "sha256", password.encode("utf8"), self._salt, 100000
     ).hex()
 
+  def make_token(self, user_id: int, icu_id: int) -> str:
+    token_hash = self.get_password_hash(json.dumps([user_id, icu_id]))
```
Comment (rth) [inline icubam/db/store.py:604]: Nice use of pbkdf2_hmac to generate a hash per user/icu pair. I would have probably tried random strings at insertion time in the DB, but there is indeed a risk if collision.

File: icubam/authenticator.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    return user_icu
+
+  def decode_from_db(self, token_str: str) -> Optional[Tuple]:
+    """Returns a User, ICU trying to get the token from the database."""
+    if len(token_str) != store.UserICUToken.TOKEN_SIZE:
+      return None
+
+    uit = self.db.get_token(token_str)
+    if uit is None:
+      logging.warning(f'No token {token_str}')
```
[the remark this answers is not in the document]
Comment (olivierteboul) [reply icubam/authenticator.py:65]: let me improve it

File: icubam/authenticator.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    """Returns a User, ICU trying read a jwt."""
+    data = self.token_encoder.decode(token_str)
+    if data is None:
+      logging.warning("No token to be found.")
+      return None
+
+    try:
+      userid, icuid = data
+    except Exception as e:
+      logging.error(f'Token is not a 2-tuple falling back to old token: {e}')
```
[the remark this answers is not in the document]
Comment (olivierteboul) [reply icubam/authenticator.py:80]: good catch!

File: icubam/authenticator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (olivierteboul) [reply icubam/authenticator.py:63]: I have switched to deciding whether to use a decoder or the other based on the token size.

**Comment by olivierteboul** [comment]
Thanks @rth for the review! Agreed i will wait for @dulacarnold 's review.

**Pull request merged**